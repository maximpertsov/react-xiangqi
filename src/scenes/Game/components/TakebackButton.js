import React, { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Button, Icon } from 'semantic-ui-react';
import requestTakeback from 'actions/requestTakeback';
import cancelTakeback from 'actions/cancelTakeback';

const TakebackButton = () => {
  const dispatch = useDispatch();

  const gameSlug = useSelector(state => state.gameSlug);
  const requestedTakeback = useSelector(state => state.requestedTakeback);
  const username = useSelector(state => state.username);

  const sendTakebackRequest = useCallback(() => {
    dispatch(requestTakeback({ gameSlug, username }));
  }, [dispatch, gameSlug, username]);

  const cancelTakebackRequest = useCallback(() => {
    dispatch(cancelTakeback({ gameSlug, username }));
  }, [dispatch, gameSlug, username]);

  const renderTakebackButton = () => (
    <Button onClick={sendTakebackRequest}>
      <Icon fitted name="undo" />
    </Button>
  );

  const renderCancelTakebackButton = () => (
    <Button
      color="red"
      icon
      labelPosition="left"
      onClick={cancelTakebackRequest}
    >
      <Icon name="undo" />
      Cancel
    </Button>
  );

  if (requestedTakeback) return renderCancelTakebackButton();
  return renderTakebackButton();
};

export default TakebackButton;
