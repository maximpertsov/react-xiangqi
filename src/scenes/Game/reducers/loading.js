export const loading = (state = false, action) => {
  switch (action.type) {
  case 'toggle_loading':
    return action.loading;
  default:
    return state;
  }
};

export default {};
