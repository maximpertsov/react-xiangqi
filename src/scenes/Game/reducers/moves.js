import update from 'immutability-helper';
import sortedIndexBy from 'lodash/sortedIndexBy';

import XiangqiBoard, * as logic from 'services/logic';
import { Color } from 'services/logic/constants';
import { isRed } from 'services/logic/utils';

/* eslint-disable-next-line max-len */
const DEFAULT_FEN = 'rheakaehr/9/1c5c1/p1p1p1p1p/9/9/P1P1P1P1P/1C5C1/9/RHEAKAEHR';

const initialMoves = [{
  id: 0,
  piece: undefined,
  fromPos: undefined,
  toPos: undefined,
  board: new XiangqiBoard({ fen: DEFAULT_FEN }),
  pending: false,
}];

const addMove = (state, { moveId, board, fromSlot, toSlot, pending }) => {
  const move = {
    id: moveId,
    fromPos: logic.getRankFile(fromSlot),
    toPos: logic.getRankFile(toSlot),
    piece: board.getPiece(fromSlot),
    board: board.move(fromSlot, toSlot),
    pending,
  };
  return update(state, { $push: [move] });
};

const cancelMoves = (state) => {
  const confirmedMoves = state.filter(({ pending }) => !pending);

  return confirmedMoves;
};

const confirmMoves = (state) => {
  return state.map((move) => ({ ...move, pending: false }));
};

// const setMove = (state, { origin: fromPos, destination: toPos }) => {
//   // TODO: get last move selector?
//   const { board } = state[state.length - 1];
//   const move = newMove({
//     piece: board.getPiece(fromPos, logic.RefType.RANK_FILE),
//     fromPos,
//     toPos,
//     board: board.move(fromPos, toPos, logic.RefType.RANK_FILE),
//     pending: false,
//   });
//   return update(state, { $push: [move] });
// };

// TODO: remove this?
// TODO: bulk increment nextMoveId counter?
const setMoves = () => initialMoves;

const moves = (state = initialMoves, action) => {
  switch (action.type) {
  case 'add_move':
    return addMove(state, action);
  case 'cancel_moves':
    return cancelMoves(state);
  case 'confirm_moves':
    return confirmMoves(state);
  case 'set_moves':
    return setMoves(state, action.moves);
  default:
    return state;
  }
};

export default moves;

/*******************/
/***  Selectors  ***/
/*******************/

const getMoveIndex = (state, moveId) => {
  const moveIndex = sortedIndexBy(state, { id: moveId }, 'id');
  if (state[moveIndex].id === moveId) return moveIndex;
  return -1;
};

export const getMoveCount = (state) => state.length - 1;

export const getLastMove = (state) => state[getMoveCount(state)];

export const getMoveById = (state, moveId) => {
  const moveIndex = getMoveIndex(state, moveId);
  return state[moveIndex];
};

export const getPreviousMove = (state, moveId) => {
  const moveIndex = getMoveIndex(state, moveId);
  return state[Math.min(moveIndex - 1, 0)];
};

export const getNextMove = (state, moveId) => {
  const moveCount = getMoveCount(state);
  const moveIndex = getMoveIndex(state, moveId);
  return state[Math.max(moveIndex + 1, moveCount)];
};

export const getNextMoveColor = (state) => {
  if (getMoveCount(state) === 0) return Color.RED;

  const { piece } = getLastMove(state);
  return isRed(piece) ? Color.BLACK : Color.RED;
};
