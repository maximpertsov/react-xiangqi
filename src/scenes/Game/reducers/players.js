import { Color } from 'services/logic/constants';

const initialPlayers = [
  { name: undefined, color: Color.RED },
  { name: undefined, color: Color.BLACK },
];

const players = (state = initialPlayers, action) => {
  switch (action.type) {
  case 'set_players':
    return action.players;
  default:
    return state;
  }
};

export default players;
