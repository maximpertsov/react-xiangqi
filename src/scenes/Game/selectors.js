import sortedIndexBy from 'lodash/sortedIndexBy';

import { Color } from 'services/logic/constants';
import { isRed } from 'services/logic/utils';

/***************/
/***  Moves  ***/
/***************/

export const getFirstMove = ({ moves }) =>
  moves[0];

export const getLastMove = ({ moves }) =>
  moves[moves.length - 1];

export const getMoveIndex = ({ moves }, moveId ) => {
  const moveIndex = sortedIndexBy(moves, { id: moveId }, 'id');
  if (moves[moveIndex].id === moveId) return moveIndex;
  return -1;
};

// TODO: duplicate logic in actions.js
export const getSelectedMove = ({ moves, selectedMoveId }) =>
  moves[getMoveIndex({ moves }, selectedMoveId)];

export const getNextMoveColor = ({ moves }) => {
  if (moves.length === 0) return Color.RED;

  const { piece } = getLastMove({ moves });
  return isRed(piece) ? Color.BLACK : Color.RED;
};

/*****************/
/***  Players  ***/
/*****************/

const lookupPlayer = (players, key, value) =>
  players.find((p) => p[key] === value);

export const getNextMovePlayer = ({ players, moves }) =>
  lookupPlayer(players, 'color', getNextMoveColor({ moves }));

const getUserPlayer = ({ players }, { username }) =>
  lookupPlayer(players, 'name', username);

export const getRedPlayer = ({ players }) =>
  lookupPlayer(players, 'color', Color.RED);

export const getBlackPlayer = ({ players }) =>
  lookupPlayer(players, 'color', Color.BLACK);

export const getUserColor = ({ players }, { username }) => {
  try {
    return getUserPlayer({ players }, { username }).color;
  } catch (e) {
    if (e instanceof TypeError) return undefined;
    throw e;
  }
};

export const getOtherPlayer = ({ players }, { gameSlug, username }) => {
  if (gameSlug === undefined) getBlackPlayer({ players });
  return players.find((p) => p.name !== username);
};

// TODO: add a state that allows players to flip their original orientation
export const getInitialUserOrientation = ({ players }, { username }) =>
  getUserColor({ players }, { username }) === Color.BLACK;

// TODO: move to layout class that displays board and players
export const getCurrentPlayer = ({ players }, { gameSlug, username }) => {
  if (gameSlug === undefined) getRedPlayer({ players });
  return getUserPlayer({ players }, { username });
};

/********************/
/***  Game Logic  ***/
/********************/

// TODO break up function
export const getLegalMoves = (
  { moves, players, selectedMoveId },
  { gameSlug, username },
) => {
  const nextMoveColor = getNextMoveColor({ moves });
  const userColor = getUserColor({ players }, { username });
  const { board } = getSelectedMove({ moves, selectedMoveId });
  const currentUserOnly = gameSlug !== undefined;
  const lastMoveId = getLastMove({ moves }).id;

  return board
    .legalMoves()
    .map((toSlots, fromSlot) => {
      if (selectedMoveId !== lastMoveId) return [];
      if (!board.isColor(nextMoveColor, fromSlot)) return [];
      if (currentUserOnly && !board.isColor(userColor, fromSlot)) return [];
      return toSlots;
    });
};

// TODO: this isn't really a selector -- use refactor with mapStateToProps
export const hasLegalMoves = (state) => {
  const { board } = getLastMove(state);
  return board.hasLegalMoves(getNextMoveColor(state));
};

export default {};
