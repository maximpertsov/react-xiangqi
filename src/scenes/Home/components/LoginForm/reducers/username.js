const username = (state = '', action) => {
  switch (action.type) {
    case 'set_login_username':
      return action.username;
    default:
      return state;
  }
};

export default username;
