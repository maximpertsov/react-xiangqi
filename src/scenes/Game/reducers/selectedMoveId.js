import { getMoveCount, getMoveIndex } from '../selectors';

const setSelectedMoveIndex = (moves, moveIndex) => {
  const { id } = moves[moveIndex];
  return id;
};

const setPreviousMove = (state, moves) => {
  const moveIndex = getMoveIndex({ moves }, state);
  return setSelectedMoveIndex(moves, Math.max(moveIndex - 1, 0));
};

const setNextMove = (state, moves) => {
  const moveIndex = getMoveIndex({ moves }, state);
  const moveCount = getMoveCount({ moves });
  return setSelectedMoveIndex(moves, Math.min(moveIndex + 1, moveCount));
};

const selectLastMove = (state, moves) => {
  const moveCount = getMoveCount({ moves });
  return setSelectedMoveIndex(moves, moveCount);
};

const selectedMoveId = (state = 0, action) => {
  switch (action.type) {
  case 'select_move':
    return action.moveId;
  case 'select_previous_move':
    return setPreviousMove(state, action.moves);
  case 'select_next_move':
    return setNextMove(state, action.moves);
  case 'select_last_move':
    return selectLastMove(state, action.moves);
  default:
    return state;
  }
};


export default selectedMoveId;
