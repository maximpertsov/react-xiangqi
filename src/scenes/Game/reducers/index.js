import { combineReducers } from 'redux';
import loading from 'scenes/Game/reducers/loading';
import moves, * as fromMoves from 'scenes/Game/reducers/moves';
import players from 'scenes/Game/reducers/players';
import selectedMoveId from 'scenes/Game/reducers/selectedMoveId';

const game = combineReducers({ loading, moves, players, selectedMoveId });

export default game;

/***************/
/***  Moves  ***/
/***************/

export const getMoveCount = ({ moves }) => fromMoves.getMoveCount(moves);

export const getLastMove = ({ moves }) => fromMoves.getLastMove(moves);

export const getSelectedMove = ({ moves, selectedMoveId }) =>
  fromMoves.getMoveById(moves, selectedMoveId);

export const getPreviousMove = ({ moves, selectedMoveId }) =>
  fromMoves.getPreviousMove(moves, selectedMoveId);

export const getNextMove = ({ moves, selectedMoveId }) =>
  fromMoves.getNextMove(moves, selectedMoveId);

export const getNextMoveColor = ({ moves }) =>
  fromMoves.getNextMoveColor(moves);
