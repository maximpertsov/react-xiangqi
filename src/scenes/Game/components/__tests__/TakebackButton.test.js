import React from 'react';
import * as selectors from 'reducers/selectors';

import takeback from 'actions/takeback';
import TakebackButton from 'scenes/Game/components/TakebackButton';

describe('TakebackButton', () => {
  let store;
  let wrapper;
  let spys = {};

  beforeEach(() => {
    setupReduxSpys(spys, store);

    spys.getCurrentPlayer = jest.spyOn(selectors, 'getCurrentPlayer');
    spys.getCurrentPlayer.mockReturnValue({ name: 'currentPlayer' });

    spys.getOpponent = jest.spyOn(selectors, 'getOpponent');
    spys.getOpponent.mockReturnValue({ name: 'opponent' });

    wrapper = shallowWrappedComponent(<TakebackButton />, store);
  });

  afterEach(() => {
    store.clearActions();
    restoreSpys(spys);
  });

  describe('default button', () => {
    beforeAll(() => {
      store = mockStore({});
    });

    beforeEach(() => {
      spys.requestTakeback = jest.spyOn(takeback, 'request');
      spys.requestTakeback.mockReturnValue(() => {});
    });

    test('snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('click', () => {
      wrapper.find('Button').simulate('click');

      expect(spys.requestTakeback).toHaveBeenCalledWith({
        gameSlug: undefined,
        username: 'currentPlayer',
      });
      expect(store.getActions()).toStrictEqual([]);
    });
  });

  describe('cancel button', () => {
    beforeAll(() => {
      store = mockStore({ openTakebackOffer: 'currentPlayer' });
    });

    beforeEach(() => {
      spys.cancelTakeback = jest.spyOn(takeback, 'cancel');
      spys.cancelTakeback.mockReturnValue(() => {});
    });

    test('snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('click', () => {
      wrapper.find('Button').simulate('click');

      expect(spys.cancelTakeback).toHaveBeenCalledWith({
        gameSlug: undefined,
        username: 'currentPlayer',
      });
      expect(store.getActions()).toStrictEqual([]);
    });
  });

  describe('accept or reject button', () => {
    beforeAll(() => {
      store = mockStore({ openTakebackOffer: 'opponent' });
    });

    beforeEach(() => {
      spys.acceptTakeback = jest.spyOn(takeback, 'accept');
      spys.acceptTakeback.mockReturnValue(() => {});
      spys.rejectTakeback = jest.spyOn(takeback, 'reject');
      spys.rejectTakeback.mockReturnValue(() => {});
    });

    test('snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('click accept', () => {
      wrapper
        .find('Button')
        .find({ color: 'green' })
        .simulate('click');

      expect(spys.acceptTakeback).toHaveBeenCalledWith({
        gameSlug: undefined,
        username: 'currentPlayer',
      });
      expect(store.getActions()).toStrictEqual([]);
    });

    test('click reject', () => {
      wrapper
        .find('Button')
        .find({ color: 'red' })
        .simulate('click');

      expect(spys.rejectTakeback).toHaveBeenCalledWith({
        gameSlug: undefined,
        username: 'currentPlayer',
      });
      expect(store.getActions()).toStrictEqual([]);
    });
  });
});
