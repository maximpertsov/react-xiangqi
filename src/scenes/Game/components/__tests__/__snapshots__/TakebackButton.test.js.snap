// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TakebackButton accept or reject button snapshot 1`] = `
<ButtonGroup>
  <Button
    as="button"
    compact={true}
  >
    <Icon
      as="i"
      name="undo"
    />
  </Button>
  <Button
    as="button"
    color="green"
    onClick={[Function]}
  >
    Accept
  </Button>
  <Button
    as="button"
    color="red"
    onClick={[Function]}
  >
    Reject
  </Button>
</ButtonGroup>
`;

exports[`TakebackButton cancel button snapshot 1`] = `
<Button
  as="button"
  color="red"
  icon={true}
  labelPosition="left"
  onClick={[Function]}
>
  <Icon
    as="i"
    name="undo"
  />
  Cancel
</Button>
`;

exports[`TakebackButton default button snapshot 1`] = `
<Button
  as="button"
  onClick={[Function]}
>
  <Icon
    as="i"
    fitted={true}
    name="undo"
  />
</Button>
`;
